packages/@sanity/form-builder/src/inputs/common/fileTarget/utils/extractFiles.ts

```diff
@@ -0,0 +1,114 @@
+/**
+ * Utilities for extracting files from dataTransfer in a predictable cross-browser fashion.
+ * Also recursively extracts files from a directory
+ * Inspired by https://github.com/component/normalized-upload
+ * Contains special handling of
+ */
+
+import {flatten} from 'lodash'
+
+export function extractPastedFiles(dataTransfer: DataTransfer): Promise<File[]> {
+  if (dataTransfer.files && dataTransfer.files.length > 0) {
+    return Promise.resolve(Array.from(dataTransfer.files || []))
+  }
+  return normalizeItems(Array.from(dataTransfer.items || [])).then(flatten)
+}
+
+export function extractDroppedFiles(dataTransfer: DataTransfer) {
+  const files: File[] = Array.from(dataTransfer.files || [])
+  const items: DataTransferItem[] = Array.from(dataTransfer.items || [])
+  if (files && files.length > 0) {
+    return Promise.resolve(files)
+  }
+  return normalizeItems(items).then(flatten)
+}
+
+function toArray<T>(v: T | null): T[] {
+  return v === null ? [] : [v]
+}
+
+function normalizeItems(items: DataTransferItem[]) {
+  return Promise.all(
+    items.map((item) => {
+      // directory
+      if (item.kind === 'file' && item.webkitGetAsEntry) {
+        let entry
+        // Edge throws
+        try {
+          entry = item.webkitGetAsEntry()
+        } catch (err) {
+          return toArray(item.getAsFile())
+        }
+        if (!entry) {
+          return []
+        }
+        return entry.isDirectory ? walk(entry) : toArray(item.getAsFile())
+      }
+
+      if (item.kind === 'file') {
+        const file = item.getAsFile()
+        return Promise.resolve(file ? [file] : [])
+      }
+
+      if (item.kind === 'string') {
+        console.warn('DataTransfer with kind="string" is currently not supported', item.kind)
+        return Promise.resolve([])
+
+        // We disabled this for now, since we don't support handling datatransfer (e.g. pasting/drag+drop) of strings.
+        // todo: Consider adding support for this later on
+
+        // note: for security reasons reading `item.type` in a non-user initiated event stack returns an empty string
+        // so the next line is there to keep it in scope
+        // const mimeType = item.type
+        // return new Promise<string>((resolve) => item.getAsString(resolve)).then((str) => {
+        //   return str ? [new File([str], 'datatransfer.txt', {type: mimeType})] : []
+        // })
```

Comment:
I propose we remove this - we have version control, after all? Maybe a small comment about what was there instead?